### PR TITLE
Create AbstractServer type for Accumulo servers

### PR DESCRIPTION
**Pull request by ctubbsii**
Create base class to help standardize the way Accumulo servers are
initialized from configuration files, launch, run, and clean up
resources during shutdown.

This establishes the base class as a concept, but over time, more
implementation that servers have in common can be put here, to make
server-side code more maintainable, and so the code in the individual
servers can be limited to the bits that make them different.

### CODE REVIEW

**Comment by ctubbsii** [comment]
Some things that could go in here in future: ServerUtil, SecurityUtil, some RPC server setup, service advertisement in ZooKeeper code, and maybe even ServerContext could be merged with this... but this is intended to be a small, incremental action... just enough so I can initialize server-side tracing in a standard way so I can upgrade HTrace to 4.1, to match Hadoop.

**Comment by ctubbsii** [comment]
All ITs passed with this, and had no issues running with uno.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java [hunk 1]
```diff
@@ -1277,14 +1280,23 @@ public void run() throws IOException, InterruptedException, KeeperException {
...
-    ServerAddress sa = TServerUtils.startServer(context, hostname, Property.MASTER_CLIENTPORT,
-        processor, "Master", "Master Client Service Handler", null, Property.MASTER_MINTHREADS,
-        Property.MASTER_THREADCHECK, Property.GENERAL_MAX_MESSAGE_SIZE);
+    ServerAddress sa;
+    try {
+      sa = TServerUtils.startServer(context, getHostname(), Property.MASTER_CLIENTPORT, processor,
+          "Master", "Master Client Service Handler", null, Property.MASTER_MINTHREADS,
+          Property.MASTER_THREADCHECK, Property.GENERAL_MAX_MESSAGE_SIZE);
+    } catch (UnknownHostException e) {
+      throw new IllegalStateException("Unable to start server", e);
```
Comment (keith-turner) [inline server/master/src/main/java/org/apache/accumulo/master/Master.java]: Could include the host+port in error message.  Sometimes this type of info is useful when on the receiving end.

Comment (keith-turner) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: All of these added error messages are nice.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java [hunk 2]
```diff
@@ -324,6 +325,8 @@ private void moveRootTabletToRootTable(IZooReaderWriter zoo) throws Exception {
   @SuppressFBWarnings(value = "DM_EXIT",
       justification = "TODO probably not the best to call System.exit here")
   private void upgradeZookeeper() {
+    ServerContext context = getContext();
```
[the remark this answers is not in the document]
Comment (ctubbsii) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: no good reason

tl;dr

I think this was just incidental to the editing I did. In a lot of places, they were using "context" to refer to the instance variable, but now needed to call getContext() to get it from the super class. To make the diff smaller, I stowed those in a local variable for reuse throughout the method. I think this one was a case where it already was using getContext() and I switched it to use the local variable instead... probably because I thought I undoing my own edit to make it use the method...without realizing it was already using the method before I edited the code.

File: server/master/src/main/java/org/apache/accumulo/master/Master.java (hunk 1 above)
Comment (ctubbsii) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: I think we only have the hostname at this point in the startup. I think the port is resolved later when the RPC services start. I'm not sure. I can look into improving this message with the hostname at least, and port if it is available.

Comment (keith-turner) [reply server/master/src/main/java/org/apache/accumulo/master/Master.java]: Hostname would be nice.   For the port I was mistaken, I saw the port property on earlier line and thought a desired port was passed in.
